contracts/FlowContractAudits.cdc

```diff
@@ -0,0 +1,215 @@
+
+            let prevAudit = FlowContractAudits.vouchers.insert(key: key, voucher)
+            if prevAudit != nil {
+                emit VoucherRemoved(key: key, recurrent: prevAudit!.recurrent, expiryBlockHeight: prevAudit!.expiryBlockHeight)
+            }
+
+            emit VoucherCreated(address: address, recurrent: recurrent, expiryBlockHeight: expiryBlockHeight, codeHash: codeHash)
+        }
+
+        // Remove a voucher with given key
+        pub fun deleteVoucher(key: String) {
+            let v = FlowContractAudits.vouchers.remove(key: key)
+            if v != nil {
+                emit VoucherRemoved(key: key, recurrent: v!.recurrent, expiryBlockHeight: v!.expiryBlockHeight)
+            }
+        }
+    }
+
+    // Used by admin to set the Auditor capability
+    pub resource interface AuditorProxyPublic {
+        pub fun setAuditorCapability(_ cap: Capability<&Auditor>)
+    }
+
+    // The auditor account will have audit access through AuditorProxy
+    // This enables the admin account to revoke access
+    // See https://docs.onflow.org/cadence/design-patterns/#capability-revocation
+    pub resource AuditorProxy: AuditorProxyPublic {
+        access(self) var auditorCapability: Capability<&Auditor>?
+
+        pub fun setAuditorCapability(_ cap: Capability<&Auditor>) {
+            self.auditorCapability = cap
+        }
+
+        pub fun addVoucher(address: Address?, recurrent: Bool, expiryOffset: UInt64?, code: String) {
+            self.auditorCapability!.borrow()!.addVoucher(address: address, recurrent: recurrent, expiryOffset: expiryOffset, code: code)
+        }
+
+        pub fun deleteVoucher(key: String) {
+            self.auditorCapability!.borrow()!.deleteVoucher(key: key)
+        }
+
+        init() {
+            self.auditorCapability = nil
+        }
+
+    }
+
+    // Can be called by anyone but needs a capability to function
+    pub fun createAuditorProxy(): @AuditorProxy {
+        return <- create AuditorProxy()
+    }
+
+    pub resource Administrator {
+        
+        // Creates new Auditor
+        pub fun createNewAuditor(): @Auditor {
+            emit AuditorCreated()
+            return <-create Auditor()
+        }
+
+        // Checks all vouchers and removes expired ones
+        pub fun cleanupExpiredVouchers() {
+            for key in FlowContractAudits.vouchers.keys {                
+                let v = FlowContractAudits.vouchers[key]!
+                if v.expiryBlockHeight != nil {
+                    if getCurrentBlock().height > v.expiryBlockHeight! {
+                        FlowContractAudits.vouchers.remove(key: key)
+                        emit VoucherRemoved(key: key, recurrent: v.recurrent, expiryBlockHeight: v.expiryBlockHeight)                
+                    }
+                }
+            }
+        }
+
+        // For testing
```

Comment:
The FVM calls `FlowContractAudits.useVoucherForDeploy` on contract deployment. This call cannot be made in the emulator since the function is scoped as `access(contract)`. As a workaround the `Administrator` resource also provides `useVoucherForDeploy` that calls `FlowContractAudits.useVoucherForDeploy`.